examples/cert-validation.cpp

```diff
@@ -0,0 +1,248 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2022 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include <iostream>
+
+#include <mococrw/x509.h>
+
+using namespace mococrw;
+
+#define CERTS "../../tests/unit/test-certs/"
+
+int main()
+{
+    X509Certificate rootCA = X509Certificate::fromPEMFile(CERTS"root3.pem");
+    X509Certificate intermediateCA = X509Certificate::fromPEMFile(CERTS"root3.int1.pem");
+    X509Certificate intermediateCA_1 = X509Certificate::fromPEMFile(CERTS"root3.int1.int11.pem");
+
+    CertificateRevocationList rootCRL = CertificateRevocationList::fromPEMFile(CERTS"root3.crl.pem");
+    CertificateRevocationList intermediateCRL = CertificateRevocationList::fromPEMFile(CERTS"root3.int1.crl_otherentry.pem");
+
+    X509Certificate cert = X509Certificate::fromPEMFile(CERTS"root3.int1.cert.pem");
+    X509Certificate cert_from_other_chain = X509Certificate::fromPEMFile(CERTS"root1.cert1.pem");
+
+    /* Certificate chain is constructed as follows:
+         _______           _______           ______
+        |       |  signs  |       |  signs  |      |
+        |root CA|-------->|int CA |-------->| cert |
+        |_______|         |_______|         |______|
+
+       Root CA is self-signed and CA's have signed CRL's.
+       These artifacts are taken from `tests` folder in the repo
+       for demonstration purposes.
+    */
+
+    /*
+        Basic certificate validation using X509Certificate::verify() against
+        issuing CA
+    */
+    {
+        std::cout << "Validating certificate against an issuing CA\n";
+        try {
+            cert.verify({intermediateCA}, {});
+        } catch (const MoCOCrWException &e) {
+            std::cerr << "Failed to validate certificate.\n";
+            std::cerr << "OpenSSL error: " << e.what() << '\n';
+            exit(EXIT_FAILURE);
+        }
+        std::cout << "Certificate successfully validated\n---\n";
+    }
+
+    /*
+        Basic certificate validation using X509Certificate::verify() against
+        issuing certificate chain
```

Comment:
Maybe be a bit more verbose here and explain the the verify method can take a vector of possible intermediate CAs as second argument. Any of those might be used to construct a valid chain. However, not all of them need to be used.